src/ir/attribute.h

```diff
@@ -0,0 +1,80 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_ATTRIBUTE_H_
+#define SRC_IR_ATTRIBUTE_H_
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/string_view.h"
+#include "src/utils/intrusive_ptr.h"
+
+namespace raksha::ir {
+
+// Base class for attributes associated with IR elements.
+class AttributeBase : public RefCounted<AttributeBase> {
+ public:
+  enum class Kind { kInt, kString };
+  AttributeBase(Kind kind) : kind_(kind) {}
+  virtual ~AttributeBase() {}
+
+  virtual Kind kind() const { return kind_; }
+  virtual std::string ToString() const = 0;
+
+ private:
+  Kind kind_;
+};
+
+using Attribute = intrusive_ptr<const AttributeBase>;
+
+class IntAttribute : public AttributeBase {
+ public:
+  static Attribute Create(int value) {
+    return Attribute(new IntAttribute(value));
+  }
+
+  int value() const { return value_; }
+
+  std::string ToString() const override {
+    return absl::StrFormat("%d", value_);
+  }
+
+ private:
+  IntAttribute(int value) : AttributeBase(Kind::kInt), value_(value) {}
+
+  int value_;
+};
+
+class StringAttribute : public AttributeBase {
+ public:
+  static Attribute Create(absl::string_view value) {
+    return Attribute(new StringAttribute(value));
+  }
+
+  absl::string_view value() const { return value_; }
+  std::string ToString() const override { return value_; }
+
+ private:
+  StringAttribute(absl::string_view value)
+      : AttributeBase(Kind::kString), value_(value) {}
+
+  std::string value_;
+};
+
+using NamedAttributeMap = absl::flat_hash_map<std::string, Attribute>;
```

Comment:
It certainly makes sense to look up `Attribute`s by name, but isn't the name part of the identity of the `Attribute`? Perhaps the string should be placed inside the `Attribute` as well as in the hash map.

Also, shouldn't the value be an `intrusive_ptr` pointing to an `Attribute` rather than a value `Attribute`?

Reply:
This map will be used within an operation to have named attributes. e.g., `ReadFieldOp`, we will have a map like this:
```
{"schema": "MyStruct", "field": "a"}
```

Reply:
Sounds good. Also, just remembered `Attribute` is an `intruisve_ptr<AttributeBase>`, so the other part of my comment doesn't matter.